### ISSUE TITLE
[Feature]: More constant symplification?

### ISSUE THREAD

**Issue by Moelf** [MilesCranmer/SymbolicRegression.jl#381]
### Feature Request
```julia
atlas_template = TemplateStructure{(:b, :e)}(
     ((; b, e), (x,)) -> b(x)^e(log(x))
)

sr_model2 = SRRegressor(
    niterations=150,
    binary_operators=[+, -, *, ^],
    unary_operators=[log],
    should_optimize_constants=true,
    seed=nothing,
    nested_constraints = [log => [log => 0, (^) => 0]],
    expression_type=TemplateExpression,
    expression_options=(; structure=atlas_template),
    loss_function = chi2_sr
)

sr_mach2 = machine(sr_model2, X, ys)

fit!(sr_mach2)
```

![image](https://github.com/user-attachments/assets/6281a9ea-af41-4cae-b43f-91d29d6db0bf)

`log(162.43)` shouldn't use `log` at all (this increase the complexity of the expression unnecessarily)

**Comment by MilesCranmer** [comment]
But in general there will be some points when the expression has redundancies like this. The genetic algorithm should eventually simplify it based on a lower complexity.

**Comment by MilesCranmer** [comment]
Also - we don't want to necessarily force simplified expressions. This is because mutations are single-step, so to get to `log(x + const)`, the genetic algorithm might prefer to go from `const` to `log(const)` to `log(x + const)`, rather than `x` to `log(x)` to `log(x + const)` which might hit a NaN. If we force simplification at every step, the `log(const)` expression wouldn't be possible.

**Comment by Moelf** [comment]
It is already set to true

**Comment by MilesCranmer** [comment]
Oh in the code you sent, it isn't set at all:

```julia
julia> print(sr_model2.should_simplify)
nothing
```

It will be set to false during the search based on the other parameters set.

Maybe you meant `should_simplify` instead of `should_optimize_constants`?

**Comment by MilesCranmer** [comment]
(Is this ok to close? Or any issues with it?)

**Comment by Moelf** [comment]
![Image](https://github.com/user-attachments/assets/fd3f0419-9656-462d-89bd-478401e052d6)

I'm seeing this re-surface (see Complexity 14 which should be further simplified into a `ax + b`)

**Comment by Moelf** [comment]
yeah, and for that matter, Complexity 8 is a simplification of Complexity 10.

But I do expect 2-nd order polynomial to show up at some point

**Comment by MilesCranmer** [comment]
SymbolicRegression.jl fundamentally doesn't "understand" what operators you pass it. To the library, those are just integers that it is swapping between and then calling evaluation on. In some ways this is annoying, because it won't automatically simplify as you point out (though it does do some very basic simplifications), but in other ways, this is nice, because it means you can input any function you want and it will be used during the evolution. Basically, SR.jl imposes no prior on operators; and no operators get special treatment. The underlying algorithm is operator agnostic.

That being said, it will still try to minimize complexity and loss. So it will naturally trend towards simpler expressions that evaluate to the same thing. Note: that simplification happens via _evolution_, not via handwritten simplification rules.

In your hall of fame, I can see that the complexity 18 expression is slightly worse in performance than the complexity 20 one. So this satisfies the rules of the dominating pareto front. An expression is only shown if it is better performance than all _simpler_ expressions. And that's what you see.

Now, sometimes it can't seem to get the same performance for an expression that should simplify to the same thing. I think this is _sometimes_ (?) due to numerical precision issues. Like how the following is true:

```julia
julia> (0.1 + 0.2) - 0.3
5.551115123125783e-17

julia> 0.1 + (0.2 - 0.3)
2.7755575615628914e-17
```

However, it could also be because of Optim.jl exiting early, and maybe it's something we could actually try to fix? Would be good to know if there's a way to prevent this sorta thing. I would also love for this type of issue to not occur.

**Comment by Moelf** [comment]
thanks for the very detailed answer, I think that covers both "why it happens" and how should we factor this into consideration for our application, we're basically something similar to your paper https://arxiv.org/abs/2411.09851 (specifically, di-jet function), but I didn't know about that paper until today!

>though it does do some very basic simplifications, though these are turned off for TemplateExpression

does this mean `should_simplify=true` has no effect? That would certainly explain what we're seeing here. On the side, Template Expression appears to only happened after that paper, maybe this is a follow up study I can do within my group.

**Comment by MilesCranmer** [comment]
Cool! Looking forward to reading it :) And yeah the different expression types were released after Ho Fung's project; my hope is that it similar types of workflows will be easier, since this puts the parametric optimization directly into the search itself.

**Comment by Moelf** [comment]
I think conceptually it should be possible to do `should_simplify` inside each component of the `Template Expression`, maybe I should open an issue to track that?

**Comment by MilesCranmer** [comment]
Sure sounds good.


Also, just for posterity, you would write the above code with the new syntax as:

```julia
atlas_template = @template_spec(expressions=(B, E)) do x
    B(x)^E(log(x))
end

sr_model = SRRegressor(;
    expression_spec=atlas_template,
)
```

Or, with parameters, like

```julia
atlas_template = @template_spec(expressions=(B, E), parameters=(p=1,)) do x
    B(x)^E(log(x)) + p[1]
end
```

**Comment by Moelf** [comment]
Ah right I guess a breaking change may be coming.

I did just now adopt the non-macro approach from the documentation:
```julia
atlas_template = TemplateStructure{(:B, :E)}(
        ((; B, E), (x,)) -> B(x)^E(log(x))
    )

...
expression_spec = TemplateExpressionSpec(; structure)
...
```

**Comment by MilesCranmer** [comment]
Oh the old syntax will still work (and continue to work for the foreseeable future). If you `@macroexpand` the macro it shows the same thing. Just a bit less typing.
